Two robots passing

Yudai-8-om opened the pull request:
## Description
basic passing

##Test
1. Run sim
2. Spread robots
3. It will pass to the best robot (based on the preliminary cost function)

Yudai-8-om committed:
initial commit

Yudai-8-om committed:
implement receiver cost

Yudai-8-om committed:
still not  working

Yudai-8-om committed:
capture done

Yudai-8-om committed:
junky pass

Yudai-8-om committed:
fix receive1

Yudai-8-om committed:
fix receive2

Yudai-8-om committed:
better receive behavior

Yudai-8-om committed:
better receive skill

Yudai-8-om committed:
update role request

Yudai-8-om committed:
cost function updated

kfu02 commented on rj_gameplay/rj_gameplay/skill/pivot.py:
```diff
@@ -33,19 +36,19 @@ def __init__(self,
         intent.motion_command.pivot_command = [pivot_command]
         intent.trigger_mode = intent.TRIGGER_MODE_STAND_DOWN
-        intent.dribbler_speed = self.dribble_speed
+        intent.dribbler_speed = float(self.dribble_speed)
```
is the cast necessary? dribble_speed is already passed in as a float.

Yudai-8-om replied:
I know, it is weird, but I got an error without float for some reason.

kfu02 commented on rj_gameplay/rj_gameplay/skill/pivot_kick.py:
```diff
@@ -47,19 +47,17 @@ def __init__(self,
-    def tick(self, 
-             robot: rc.Robot, 
-             world_state: rc.WorldState, 
+    def tick(self, robot: rc.Robot, world_state: rc.WorldState,
```
add a comment here that describes the flow (which I believe from reading goes capture -> pivot -> kick?) since the if-else order doesn't match the sequence of actions

kfu02 commented on rj_gameplay/rj_gameplay/skill/receive.py:
```diff
@@ -24,26 +24,23 @@
         if self.settle.is_done(world_state):
-            return self.capture.tick(robot, world_state)
+            return self.capture.tick(robot, world_state, intent)
+        # elif np.linalg.norm(world_state.ball.vel) > RECEIVE_BALL_SPEED_THRESHOLD:
```
is this intentionally commented out? if it's not being used I would advise deleting the THRESHOLD param

kfu02 commented on rj_gameplay/rj_gameplay/skill/settle.py:
```diff
@@ -5,48 +5,66 @@
         if self.robot is None:
             return False
-        if world_state.our_robots[self.robot.id].has_ball_sense or np.linalg.norm(world_state.ball.vel) < SETTLE_BALL_SPEED_THRESHOLD:
+        if world_state.our_robots[self.robot.id].has_ball_sense or (
```
could you put a comment explaining what this if statement checks? it looks like "if ball sense OR (ball closer to me than passer AND ball is slow)" but that took some time to figure out

kfu02 commented on rj_gameplay/stp/role/assignment/naive.py:
```diff
@@ -109,11 +112,9 @@ def compute_costs_matrix(
-                        )
-                    )
+                raise ValueError(
+                    "Got a non-finite cost ({}) for request {} and unassinged robot"
```
```suggestion
                    "Got a non-finite cost ({}) for request {} and unassigned robot"
```

kfu02 commented on rj_gameplay/stp/role/assignment/naive.py:
```diff
@@ -71,28 +74,28 @@ def compute_costs_matrix(
             for robot_idx, robot in enumerate(free_robots):
 
                 # If the robot is not visible then do not consider its cost
+                # Cannot be infinte, will break linear_sum_assignment
```
```suggestion
                # Using 1e9 because float('inf') will break linear_sum_assignment
```

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -106,44 +106,98 @@ def __call__(
 
         if robot is None or self.target_point is None:
-            return 99
+            return 1e9
```
it would be nice if we made role assignment auto-cast float('inf') to 1e9 in the role assigner so we didn't have to remember this magic number

kfu02 replied:
added as new task to board, dw about it

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -181,14 +234,25 @@ def create_request(self, **kwargs) -> role.RoleRequest:
         pass
 
     def find_potential_receiver(self, world_state: rc.WorldState) -> rc.Robot:
```
any reason this one is named "find_potential_receiver" and the one below is "find_passer"? (why "potential"?)

kfu02 replied:
also, this doesn't have to be fixed for this PR, but these funcs are so similar that they could be one "find min" function that takes in a cost func as a param: https://www.geeksforgeeks.org/passing-function-as-an-argument-in-python/

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -106,44 +106,98 @@ def __call__(
+        angle_threshold = 5
+        dist_threshold = 1.5
+        backpass_punish_weight = 0.5
+        if robot.id != self.passer_robot.id:
```
isn't this if already covered in line 164 above?

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -106,44 +106,98 @@ def __call__(
+                                       robot.pose[0:2])
+
+            cost = 0
+            goal_to_receiver = np.linalg.norm(robot.pose[0:2] -
```
add a few comments explaining chunks of below code, e.g. here I would put 
"add to cost for backwards passes (scaled by backpass_punish_weight)"

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -106,44 +106,98 @@ def __call__(
+        # if self.chosen_receiver is not None and self.chosen_receiver.id == robot.id:
+        #     return -1e9
 
         # TODO: pick "most open" pass
```
this TODO is done, right?

kfu02 commented on rj_gameplay/rj_gameplay/tactic/pass_tactic.py:
```diff
@@ -222,18 +283,27 @@ def tick(self, world_state: rc.WorldState,
+            self.pivot_kick.skill.pivot_point = np.array(
+                pivot_result[0].role.robot.pose[0:2])
+
+            # self.receiver_cost.potential_receiver = receive_result[
```
delete this commented out code if it's not being used, or leave comment explaining why it should stay

kfu02 reviewed (changes requested):
This is great, way better than what we've had so far. A few things:
 - when I put the ball in the top left and hit start, sim crashes often, giving this error: "[planner_node-7] terminate called after throwing an instance of 'std::invalid_argument'
[planner_node-7]   what():  Cannot splice trajectories a and b, where a.last() != b.first()" No idea what this means. I'm fine with solving this in a later PR.
 - when the pass doesn't crash sim (i.e. ball starts in top right) the receiver often passes it right back to the original passer. This issue should be fixed for this PR. My hypothesis is that the pass tactic doesn't return done because neither robot has captured the ball yet, so it reassigns the new closest robot to the ball as the passer and the old passer as the new receiver and does the pass again.
 - I left a few comments in the files. Most of them are just "please add explanatory comment here". Will make future dev easier.
Nice work, @Yudai-8-om @adithya-gv and @HussainGynai

Yudai-8-om committed:
Update rj_gameplay/stp/role/assignment/naive.py

Co-authored-by: Kevin Fu <30275369+kfu02@users.noreply.github.com>

Yudai-8-om committed:
Update rj_gameplay/stp/role/assignment/naive.py

Co-authored-by: Kevin Fu <30275369+kfu02@users.noreply.github.com>

Yudai-8-om committed:
capture fail rn

Commit:
attempt to fix trigger-on-break-beam bug

Commit:
change dribble_speed to dribbler_speed to match cpp

Commit:
fix kick div by 0 bug, add TODO for settle param in line_kick

Commit:
change receive to use intercept -> capture instead of settle -> capture, write move-based intercept instead of planner-based

Commit:
bugfix: pivot_kick now updates target/pivot every tick

Commit:
rm debug prints

Commit:
rm outdated receive todo

Commit:
fix robot None bug

Commit:
implement basic passing

 - buggy bc doesn't end after one pass (aka Receive never ends)
 - implemented as behavior tree and by circumventing RoleAssignment:
 - passer/receiver are picked on first get_requests() call
 - role costs are set to always pick same passer/receiver
 - first, passer is requested
 - when passer finishes PivotKick, receiver is requested
 - receiver then goes to ball (in theory, poorly implemented rn)
 - when receiver gets ball, Tactic should end (again in theory)

kfu02 commented:
I think this is a pretty perfect example as to how our gameplay system is unnecessarily over-complicated. See my last commit msg:  
```
implement basic passing

 - buggy bc doesn't end after one pass (aka Receive never ends)
 - implemented as behavior tree and by circumventing RoleAssignment:
 - passer/receiver are picked on first get_requests() call
 - role costs are set to always pick same passer/receiver
 - first, passer is requested
 - when passer finishes PivotKick, receiver is requested
 - receiver then goes to ball (in theory, poorly implemented rn)
 - when receiver gets ball, Tactic should end (again in theory)
```

kfu02 committed:
rm debug prints in pass tactic

kfu02 committed:
fake intercept planner in intercept skill

kfu02 committed:
debug prints
